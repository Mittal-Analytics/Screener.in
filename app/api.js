"use strict";
import Rest from 'fetch-on-rest'

function getCookie(sKey) {
  if (!sKey) { return ''; }
  return decodeURIComponent(document.cookie.replace(new RegExp("(?:(?:^|.*;)\\s*" + encodeURIComponent(sKey).replace(/[\-\.\+\*]/g, "\\$&") + "\\s*\\=\\s*([^;]*).*$)|^.*$"), "$1")) || '';
}


function addOptions(defaults, url) {
  var endsWith = function(haystack, needle) {
    return haystack.indexOf(needle, haystack.length - needle.length) !== -1;
  };
  defaults.credentials = 'same-origin';
  if(defaults.method == 'raw')
    defaults.headers.Accept = 'text/html,*/*';
  if(endsWith(url, '.html'))
    defaults.headers.Accept = 'text/html,*/*';

  if(defaults.method != 'get')
    defaults.headers['X-CSRFToken'] = getCookie('csrftoken');
}

const useTrailingSlashes = true;
var Api = new Rest('/api', addOptions, useTrailingSlashes);

Api.getCsrf = function() {
  return getCookie('csrftoken');
};

Api.me = ['users', 'me'];
Api.search = ['company', 'search'];
Api.ratioSearch = ['ratios', 'search'];

Api.my = function(resource) {
  return ['users', resource];
};

Api.logout = function() {
  var url = '/logout/';
  var options = this._getOptions(undefined, 'post', url);
  var raw = window.fetch(url, options);
  return raw.then(this.parseJson);
};

Api.company = function(exchange_code, consolidated) {
  var paths = ['company', exchange_code];
  if (consolidated)
    paths.push('consolidated');
  return paths;
};

Api.cid = function(cid, component) {
  return ['company', cid, component];
};

export default Api
