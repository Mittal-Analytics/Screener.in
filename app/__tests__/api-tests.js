"use strict";
jest.mock('fetch-on-rest');
import api from '../api.js'

describe('test REST apis', function () {

  beforeEach(() => {
    window.fetch.mockClear()
  })

  afterEach(() => {
    expect(api.getPending()).toEqual([]);
  });

  it('calls the post api', function() {
    api.setResponse('/logout/', JSON.stringify({}));
    return api.logout().then(() => {
      expect(window.fetch.mock.calls).toMatchSnapshot()
    });
  });

  it('calls the get api', function() {
    api.setResponse('/api/users/me/', JSON.stringify({foo: 'bar'}));
    return api.get(api.me).then(resp => {
      expect(resp).toEqual({foo: 'bar'});
      expect(window.fetch.mock.calls).toMatchSnapshot()
    })
  });

  it('calls the delete api', function() {
    api.setResponse('/api/screens/33/?foo=bar', "{}");
    return api.delete(['screens', 33], {foo: 'bar'}).then(() => {
      expect(window.fetch).toBeCalledWith(
        '/api/screens/33/?foo=bar',
        { credentials: 'same-origin',
          headers: {
            Accept: 'application/json',
            'X-CSRFToken': ''
          },
          method: 'delete'
        }
      );
    });
  });

});


describe('test dependent libraries', function(){
  it('checks expansions', function() {
    expect(api._getUrl(api.me)).toEqual('/api/users/me/');
    expect(api._getUrl(api.me, {})).toEqual('/api/users/me/');
    expect(api._getUrl(api.me, {foo: 'bar'})).toEqual('/api/users/me/?foo=bar');
  });
});
