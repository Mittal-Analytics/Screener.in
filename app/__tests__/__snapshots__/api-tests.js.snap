exports[`test REST apis calls the get api 1`] = `
Array [
  Array [
    "/api/users/me/",
    Object {
      "credentials": "same-origin",
      "headers": Object {
        "Accept": "application/json",
      },
      "method": "get",
    },
  ],
]
`;

exports[`test REST apis calls the post api 1`] = `
Array [
  Array [
    "/logout/",
    Object {
      "credentials": "same-origin",
      "headers": Object {
        "Accept": "application/json",
        "X-CSRFToken": "",
      },
      "method": "post",
    },
  ],
]
`;
